language-support/ts/codegen/src/TsCodeGenMain.hs

```diff
@@ -107,31 +111,75 @@ genModule curPkgId mod
     _: _: _: _ -> error "IMPOSSIBLE: multi-part type constructor of more than two names"
-    [conName] -> genDefDataType curPkgId conName curModName tpls def
-    [c1, c2] -> ((makeNamespace $ map ("  " <>) typs, makeNamespace $ map ("  " <>) sers), refs)
+
+    [conName] -> genDefDataType curPkgId conName mod tpls def
+    [c1, c2] -> ((makeNamespace $ map ("  " <>) typs, []), refs)
       where
-        ((typs, sers), refs) = genDefDataType curPkgId c2 curModName tpls def
-        ns = c1 <> "_NS"
+        ((typs, _), refs) = genDefDataType curPkgId c2 mod tpls def
+        ns = c1
         makeNamespace stuff =
           [ "// eslint-disable-next-line @typescript-eslint/no-namespace"
-          , "export namespace " <> ns <> " {"] ++ stuff ++ ["} //namespace " <> ns] ++ [""]
+          , "export namespace " <> ns <> " {"] ++ stuff ++ ["} //namespace " <> ns]
 
-genDefDataType :: PackageId -> T.Text -> ModuleName -> NM.NameMap Template -> DefDataType -> (([T.Text], [T.Text]), Set.Set ModuleRef)
-genDefDataType curPkgId conName curModName tpls def =
+genDefDataType :: PackageId -> T.Text -> Module -> NM.NameMap Template -> DefDataType -> (([T.Text], [T.Text]), Set.Set ModuleRef)
+genDefDataType curPkgId conName mod tpls def =
     case dataCons def of
         DataVariant bs ->
           let
             (typs, sers) = unzip $ map genBranch bs
-            typeDesc = [""] ++ typs
-            serDesc  = ["() => jtv.oneOf<" <> conName <> typeParams <> ">("] ++ sers ++ [")"]
-          in
-          ((makeType typeDesc, makeSer serDesc), Set.unions $ map (Set.setOf typeModuleRef . snd) bs)
+            typeDesc = makeType ([""] ++ typs)
+            typ = conName <> typeParams -- Type of the variant.
+            assocDefs = assocDefDataTypes mod conName -- Associated types.
+            serDesc =
+              if not $ null paramNames -- Polymorphic type.
+              then -- Companion function.
+                let
+                  -- Any associated serializers.
+                  assocSers = map
+                    (\(n, d) -> serFromDef id n curPkgId mod tpls d)
+                    assocDefs
+                  -- The variant deserializer.
+                  function = (makeSer ( ["() => jtv.oneOf<" <> typ <> ">("] ++ sers ++ [")"]) & onLast (<> ";"));
+                  props = -- Fix the first and last line of each serializer.
+                    concatMap (onHead (erase "export const ") . onLast (<> ";")) assocSers
+                  -- The complete definition of the companion function.
+                  in function ++ props
+                     -- To-do: Can we formulate a static implements
+                     -- serializable check that works for companion
+                     -- functions?
+              else -- Companion object.
+                let
+                  assocNames = map (snd . fst) assocDefs
+                  -- Any associated serializers, dropping the first line
```

Comment:
How does the first line look like? Can we modify it with `onHead` instead?

Reply:
I recall evaluating that and determining that it was easier to discard the line.